### PR TITLE
Storage, Container and Product structures

### LINKED ISSUE #3: Implement classes for storage,  containers and product

### PR DESCRIPTION
**Pull request by N-Wouda**
1. Storage ('Warehouse', 'Factory', or 'Store'). A place where goods are stored for long-term.
2. Container ('Rack', 'Barrel', 'Box'). A storage box within a Storage facility. Contains products.
3. Product. Has properties (perishability? Price? Size?), and is contained within Containers.

Feel free to comment and discuss. Once finished, this closes #3.

### CODE REVIEW

File: classes/product.js [hunk 1]
```diff
@@ -0,0 +1,41 @@
+class Product
+{
+    /**
+     * @constructor Represents a product.
+     *
+     * @param {string} name - The display name for this product.
+     * @param {int} quantity - The amount of this product
+     * @param {int} perishable - The number of days a perishable good remains fresh. 0 for infinite.
+     */
+    constructor(name, quantity, perishable = 0)
```
Comment (N-Wouda) [inline classes/product.js]: We should do something with product dimensions (else the Container cannot know its maximum contents). Could be weight, size, volume? Any ideas?

Comment (lukko) [reply classes/product.js]: Maybe have like a generic measurement in 'units'.
For example: a container has a capacity of 100 units. A bike equals 5 units and a chair equals 2 units, then one container can contain:
- 20 bikes or
- 16 bikes and 10 chairs

Has the advantage that we don't have to deal with length, width and height of products for now.

Comment (N-Wouda) [reply classes/product.js]: Seems reasonable to me

File: classes/product.js [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+     * @param {string} name - The display name for this product.
+     * @param {int} quantity - The amount of this product
+     * @param {float} size - A unitless measure for the size of a product (qty = 1).
+     * @param {int=1} perishable - The number of days a perishable good remains fresh. 0 for infinite.
+     */
+    constructor(name, quantity, size, perishable = 0)
+    {
+        this.name = String(name);
+        this.quantity = parseInt(quantity);
+        this.size = parseFloat(size);
```
Comment (N-Wouda) [inline classes/product.js]: @lukko Does this seem reasonable? I think this is pretty much what you suggested :).

File: classes/container.js [hunk 3]
```diff
@@ -0,0 +1,48 @@
+class Container
```
Comment (N-Wouda) [inline classes/container.js]: TODO: A way to remove products from a container.

Have this be done automatically when the quantity on any product drops to zero (could be done when calculating the total used capacity).

File: classes/storage.js [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+    /**
+     * @param {*} item - Any item to be put in storage. Type checking is delegated to subclasses.
+     * @throws Error when capacity would be exceeded by adding this item to the Storage!
+     */
+    addItem(item)
+    {
+        var availableCapacity = this.capacity - this._usedCapacity();
+
+        if (!availableCapacity) {
```
Comment (N-Wouda) [inline classes/storage.js]: Should check for `<= 0`, not just zero.

**Comment by N-Wouda** [comment]
@JanVeen @VinnieBoySlim I do not have write access to the repo, but I think this is part is finished for now. Feel free to review and if suitable, merge the PR.

Also, could I perhaps be given write access so I can merge PRs myself? :)

**Comment by VincentVelthuizen** [comment]
@N-Wouda Since I think it is appropriate for you to have elevated rights I have just assigned you admin privilege on this repo. The default for the team is still read, we could make it write, what do you think?

**Comment by N-Wouda** [comment]
> The default for the team is still read, we could make it write, what do you think?

'Write' allows merging pull requests, which may not be desirable. Not that I think anyone would intentionally push faulty code, but accidents happen. Read should suffice for opening PRs. Then merging can be done by anyone with Write access.

**Comment by N-Wouda** [comment]
If no further comment/review I am merging this later today.

**Comment by lukko** [comment]
Why don't you use camelcase for file names?

**Comment by N-Wouda** [comment]
> Why don't you use camelcase for file names?

I generally follow something along the lines of [this SO answer](http://stackoverflow.com/questions/7273316/what-is-the-javascript-filename-naming-convention#answer-7273448) for JS code, but more generally, [this also holds true](http://stackoverflow.com/questions/13511227/url-structure-lower-case-vs-upper-case):

> The path part is or is not case sensitive, depending on the server environment and server. Typically Windows machines are case insensitive, while Linux machines are case sensitive. This means that you should stick to lowercase or you risk introducing a bug that's really hard to hunt down (mismatched case that doesn't matter on the dev server).

**Pull request merged**